---
title: "Analysis of RNAseq from Severely Malnourished and non-Severely Malnourished Individuals"
author: "Arthur J. VanValkenburg"
date: "4/24/2019"
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float: true
    theme: "flatly"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
suppressPackageStartupMessages({
  library(SummarizedExperiment)
  library(edgeR)
  library(sva)
  library(SingleCellExperiment)
  library(singleCellTK)
  library(DESeq2)
  library(TBSignatureProfiler)
  library(DT)
  library(enrichR)
  library(Rtsne)
  library(umap)
  library(ggplot2)
  library(ComplexHeatmap)
  library(tidyverse)
  library(stringi)
  library(hypeR)
  library(knitr)
  library(kableExtra)
  
})
  
knitr::opts_chunk$set(cache = TRUE)
knitr::opts_chunk$set(echo = FALSE)

#  library(reshape2)
#  library(RColorBrewer)
#  library(limma)
#  library(knitr)
#  library(pander)
#  library(xtable)
#})
```

# Loading and processing data
```{r load data, include=FALSE}

#indata is a SingleCellExperiment of the combined datasets

indata <- readRDS("new_indata_SCE.rds")

## combined datasets, used to make "new_indata_SCE.rds". Code included for reference. 
#indatalist <- SummarizedExperiment(assays = SimpleList(counts = as.matrix(indatalist$indata2)), colData = DataFrame(indatalist$metadata_added))
#metadata_added = indatalist$metadata_added
#rm(indatalist)

# Want to have 5% present rate
indata <- indata[apply(assay(indata,"counts") != 0, 1, mean)>.2,] 

## Correct some errors in metadata
colData(indata)$Tb_status[grep("074", colnames(indata))] <- "LTBI"
colData(indata)$Tb_status[grep("082", colnames(indata))] <- "LTBI"
colData(indata)$Tb_status[grep("111", colnames(indata))] <- "LTBI"

## Make CPM and log_cpm
#indata  = mkAssay(indata, log = TRUE, counts_to_CPM = TRUE)
#assays(indata)

## Make Severely Malnourished column data
colData(indata)$Tb_status <- factor(gsub("activeTB", "TB", as.character(colData(indata)$Tb_status)))
colData(indata)$bmi_cat2 <- factor(gsub("severely malnourished", "mal", gsub("well nourished", "well", colData(indata)$bmi_cat2)))
colData(indata)$status <- factor(paste(colData(indata)$Tb_status, colData(indata)$bmi_cat2, sep="_"))
table(colData(indata)$status)


## Removing Outliers
indata <- indata[,-which(colnames(indata) %in% c("10200459B","10200335A","10200332A"))]

```



### ComBat Batch correction
```{r corrected data, message=FALSE, results='hide', include = FALSE}

## Make batch variable
batch=rep(1,ncol(indata))
batch[colnames(indata) %in% c("10200247B")]=1
batch[colnames(indata) %in% c("10200227B","10200346A","10200381A","10200308A","10200361A","10200365B","10200467B","10200341A","10200374A", "10200548B","10200343A","10200374B","10200346A","10200377A") ]=1
batch[colnames(indata)%in% c("10200070A", "10200128A","10200056A", "10200075A","10200136A", "10200105A","10200034A","10200058A", "10200007A", "10200064A","10200159A","10200077A","10200073A","10200130A","10200181A","10200059A","10200040B","10200018B","10200065B","10200003B","10200024B","10200004B","10200009B","10200001B","10200057B","10200029B","10200077B","10200035B","10200061B","10200030B","10200063B","10200034B","10200010A","10200288A","10200082A","10200074A","10200060A","10200023A","10200140A","10200111A","10200112A","10200016A","10200015A","10200087A")]=2
colData(indata)$batch <- factor(batch)
rm(batch)

modcombat <- model.matrix(~colData(indata)$bmi_cat2+colData(indata)$Tb_status)

combined_India.combatSeqCorrect <- ComBat_seq(assay(indata, "counts"),
                                      colData(indata)$batch, group=NULL,
                                      covar_mod=modcombat)

assay(indata, "combatseq") <- combined_India.combatSeqCorrect
assays(indata)
indata <- mkAssay(indata, input_name = "combatseq", log = TRUE)


rm(modcombat, combined_India.combatSeqCorrect)

```




### Keep only LTBI
```{r filter, include=FALSE}
indata = indata[,colData(indata)$Tb_status=="LTBI"]
dim(indata)
table(colData(indata)$status)
 
```

### Keep only protein coding genes
```{r filter-by-gene-type, include=FALSE}

#indata <- as(indata, "SingleCellExperiment")

prot_code <- indata@rowRanges@elementMetadata
prot_ <- prot_code[which(prot_code$V2 == "protein_coding"),]
prot <- as.character.Array(prot_@listData$V1)

indata = indata[indata@rowRanges@partitioning@NAMES %in% prot,]

rm(prot_, prot_code, prot)
```


### Make SCE for SCTK {.unlisted, .unnumbered}
```{r sce-for-sctk, echo=FALSE}
indata_SCE = as(indata, "SingleCellExperiment")
#singleCellTK(indata_SCE)
saveRDS(indata_SCE, file = "indata_SCE.rds")
```


# Dimension Reduction Plots {.tabset}

## PCA  {.tabset}
Using the DESeq object and PCAplot function on the ComBat corrected data


```{r PCA, fig.cap="PCA plot created with the DESeq object and PCAplot function on the Combat-Seq batch corrected data.",  echo=FALSE}

# Changed plotPCA function to plot PCA2 and PCA3
plotARTPCA <- function (object, intgroup = "condition", ntop = 500, returnData = FALSE) 
{
    rv <- rowVars(assay(object))
    select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, 
        length(rv)))]
    pca <- prcomp(t(assay(object)[select, ]))
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    if (!all(intgroup %in% names(colData(object)))) {
        stop("the argument 'intgroup' should specify columns of colData(dds)")
    }
    intgroup.df <- as.data.frame(colData(object)[, intgroup, 
        drop = FALSE])
    group <- if (length(intgroup) > 1) {
        factor(apply(intgroup.df, 1, paste, collapse = ":"))
    }
    else {
        colData(object)[[intgroup]]
    }
    d <- data.frame(PC2 = pca$x[, 2], PC3 = pca$x[, 3], group = group, 
        intgroup.df, name = colnames(object))
    if (returnData) {
        attr(d, "percentVar") <- percentVar[2:3]
        return(d)
    }
    ggplot(data = d, aes_string(x = "PC2", y = "PC3", 
        color = "group")) + geom_point(size = 3) + xlab(paste0("PC2: ", 
        round(percentVar[2] * 100), "% variance")) + ylab(paste0("PC3: ", 
        round(percentVar[3] * 100), "% variance")) + coord_fixed()
}

indata_tmp <- SummarizedExperiment(assays=list(counts=assays(indata)$log_cpm),
                     colData = colData(indata))

plotARTPCA( DESeqTransform(indata_tmp), intgroup = "BMI")

```

## tSNE {.tabset}


```{r tSNE, fig.cap="tSNE", fig.height=4, fig.width=4,  echo=FALSE}
assay_type = "log_cpm"

set.seed(1)
tsne_out <- Rtsne(t(assay(indata,assay_type)), check_duplicates = FALSE, pca = TRUE, perplexity=10, theta=0.5, dims=2)

embedding <- as.data.frame(tsne_out$Y)
embedding$Class <- (indata$BMI )


g <- ggplot(embedding, 
            aes(x=V1, y=V2, color=Class), label = colnames(assay(indata,assay_type))) + geom_point(size=1.5) + xlab("T-SNE 1") + ylab("T-SNE 2") + theme(plot.title = element_text(hjust = 0.5)) + 
ggtitle("TSNE Plot") + scale_color_gradient(low="blue", high="red")

plot(g)
```

## UMAP {.tabset}


```{r UMAP, fig.cap="UMAP", fig.height=4, fig.width=4, echo=FALSE}
assay_type = "log_cpm"

set.seed(1)
umap_out <- umap(t(assay(indata,assay_type)))

embedding <- as.data.frame(umap_out$layout)
embedding$Class <- as.factor(indata$bmi_cat2)

g <- ggplot(embedding, aes(x=V1, y=V2, color=Class, label = colnames(assay(indata,assay_type)))) + geom_point(size=1.5) + xlab("UMAP 1") + ylab("UMAP 2") + theme(plot.title = element_text(hjust = 0.5)) + 
ggtitle("UMAP Plot")

plot(g)
```


# Differential expression {.tabset}
```{r diffex}
## make a matrix for differential expression
designMat <- model.matrix(~factor(bmi_cat2), data=colData(indata))

colnames(designMat)
colnames(designMat) <- c("Intercept", "wellnourished")

head(designMat)

fit <- lmFit(assay(indata, "log_cpm"), designMat)

## Difference in expression between well-nourished and malnourished individuals with LTBI individuals, among malnourished individuals

contrast.matrixNut<- makeContrasts(wellnourished, levels = designMat)
fitNut <- contrasts.fit(fit,contrast.matrixNut)
fitNut <- eBayes(fitNut)
limmaResNut <- topTable(fitNut, adjust.method = "BH", n = Inf, sort.by = "P")


dim(limmaResNut[limmaResNut$adj.P.Val <0.05,])
table(colData(indata)$bmi_cat2)

## histograms to analyze adjusted p-values and p-values of limma results; shows spike near 0
new_hist_padjval_100breaks <- hist(limmaResNut$adj.P.Val, breaks=100)
new_hist_pval_100breaks <- hist(limmaResNut$P.Value, breaks=100)

# keep FDR rate < 0.05 and order by FDR for limma results
sig_limmaResNut <- limmaResNut[limmaResNut$adj.P.Val <0.05,]
sig_limmaResNut <- sig_limmaResNut[order(sig_limmaResNut$adj.P.Val),]

#write.csv(sig_limmaResNut, "sig_TBgenes_padj.csv")

## make top 500 gene list
top_500_genes <- limmaResNut[1:500,]


```



## All genes datatable and heatmap {.tabset}
```{r diffex-datatable-heatmap}
datatable(top_500_genes, 
          options=list(scrollX=T,pageLength=20),
          rownames = T)

## Make a Heatmap of top 500 genes
mat = as.matrix(assay(indata[row.names(top_500_genes),],"log_cpm"))
mat = t(scale(t(mat)))
mat = mat[,order(colData(indata)$"BMI")]

## df for heatmap annotation
df=data.frame(Nutrition=colData(indata)$"bmi_cat2", BMI=(colData(indata)$"BMI")) 

df=df[order(df$BMI),]

## heatmap annotation object
ha = HeatmapAnnotation(df = df, col = list(TB_status=c("mal"="Red","well"="Blue")))

## heatmap ordered by BMI
Heatmap(mat,show_row_names=F,show_column_names = T, top_annotation = ha, cluster_columns = FALSE )
```


# SignatureProfiler {.tabset}

## TBSigProfiler Pathways {.tabset}
```{r, message=FALSE, results='hide'}


## Add new signatures to the profiler *later releases of TBSignatureProfiler may not require this
TBsignatures$Zhao_NANO_6 <- c("ANKRD22","ATP1A1","BLK","CCR6","DARS2","EXOC2")

TBsignatures$'CYTO6' <- c("CCL11", "IFNG", "IL15", "IL1B", "IL6", "CXCL10")

TBsignatures$Leong_RISK_29 <- c("SRBD1", "ZNF419", "SH2D1B","CTSA", "GSTA4", "AGAP9", "MOB3C", "WARS1", "LUC7L", "ZNRF1", "CIRBP", "PRSS53", "APOL6", "TCN2", "MDN1", "SNRNP70", "SLC3A1", "NAGA", "SPDYE5",  "SPSB1", "CCDC14",  "IL31RA", "DERA", "FUT4",
"NEIL1",   "ENO3",   "CCDC78", "HM13","ZNF202" )

TBsignatures$Zak_RISK_16 <- c("ANKRD22","APOL1","BATF2" ,"ETV7","FCGR1A", "FCGR1B",  "GBP1","GBP2","GBP4","GBP5", "SCARF1","SEPTIN4", "SERPING1", "STAT1","TAP1","TRAFD1")
TBsignatures$Suliman_RISK_4 <- c("GAS6", "SEPTIN4", "CD1C", "BLK")

## selected tb signatures
samp_tbsignatures <- list(TBsignatures$Suliman_RISK_4, TBsignatures$Sweeney_OD_3, TBsignatures$Leong_RISK_29, TBsignatures$Zak_RISK_16)
names(samp_tbsignatures) <- c("Suliman_RISK_4", "Sweeney_OD_3", "Leong_RISK_29", "Zak_RISK_16")

gsva_res <- runTBsigProfiler(indata, useAssay = "log_cpm", algorithm = "GSVA",
                             signatures = samp_tbsignatures, parallel.sz = 4)
ssgsea_res <- runTBsigProfiler(indata, useAssay = "log_cpm", algorithm = "ssGSEA",
                               signatures = samp_tbsignatures, parallel.sz = 4)


```

### GSVA {.tabset}

#### Heatmap

```{r subgsva_a_TBsigs}
signatureHeatmap(gsva_res, name="GSVA", signatureColNames = names(samp_tbsignatures),
                 annotationColNames = c("bmi_cat2"),
                 showColumnNames = FALSE,
                 split_heatmap='none')
```


#### Boxplot

```{r boxgsva_TBsigs}
signatureBoxplot(gsva_res, name="GSVA", signatureColNames = names(samp_tbsignatures),
                 annotationColName = c("bmi_cat2"))# , rotateLabels = TRUE)
```

#### Boxplots Single {.tabset}

```{r , results="asis"}
for (i in names(samp_tbsignatures)){

  cat("#####", i, "\n")

  print(signatureBoxplot(gsva_res,
                         name=i, 
                         signatureColNames = i,
                         annotationColName = c("bmi_cat2"),   
                         violinPlot = TRUE,
                         rotateLabels = T))

  cat("\n\n")
}

```

#### Signature plots {.tabset}
```{r genes_gsva_TBsigs, results="asis"}
for (i in names(samp_tbsignatures)){

  cat("#####", i, "\n")

  signatureGeneHeatmap(gsva_res, useAssay="log_cpm",
                     samp_tbsignatures[[i]],
                     name = i, signatureColNames = NULL,
                     annotationColNames = c("bmi_cat2",i),
                     showColumnNames = TRUE, 
                     column_order = ordered_bmi[,1])

  cat("\n\n")
}

```

#### AUC Table {.tabset}
```{r, message = FALSE}
set.seed(0)
tableAUC(gsva_res,
         annotationColName = "bmi_cat2",
         signatureColNames = names(samp_tbsignatures),
         num.boot = 100,
         pb.show = FALSE)
```

#### AUC Boxplots
```{r, message = FALSE}
set.seed(0)
compareBoxplots(gsva_res, annotationColName = "bmi_cat2",
                signatureColNames = names(samp_tbsignatures),
                pb.show = FALSE, fill.col = "blue",
                rotateLabels = TRUE)
```

#### ROC plots
```{r, message = FALSE, fig.height = 9, fig.width = 12}
signatureROCplot_CI(inputData = gsva_res,
                   signatureColNames = names(samp_tbsignatures),
                   annotationColName = "bmi_cat2")

```

#### Separate ROC plots  {.tabset}

```{r, results = 'asis', message = FALSE}
for (i in names(samp_tbsignatures)){

  cat("#####", i, "\n")

  print(signatureROCplot_CI(inputData = gsva_res,
                   signatureColNames = i,
                   annotationColName = "bmi_cat2",
                   name = paste("ROC plot,", i, sep = " ")))

  cat("\n\n")
}
```


### ssGSEA {.tabset}

#### Heatmap

```{r subssgsea_a_TBsigs}
signatureHeatmap(ssgsea_res, name="ssGSEA", signatureColNames = names(samp_tbsignatures),
                 annotationColNames = c("bmi_cat2"),
                 showColumnNames = FALSE, scale = TRUE,
                 split_heatmap='none')
```

#### Boxplot

```{r boxssgsea_TBsigs}
signatureBoxplot(ssgsea_res, name="ssGSEA", signatureColNames = names(samp_tbsignatures),
                 annotationColName = c("bmi_cat2"), scale = TRUE) #rotateLabels = TRUE,
```

#### Boxplots Single {.tabset}

```{r , results="asis"}
for (i in names(samp_tbsignatures)){

  cat("#####", i, "\n")

  print(signatureBoxplot(ssgsea_res, name=i, signatureColNames = i,
                 annotationColName = c("bmi_cat2"), rotateLabels = T))

  cat("\n\n")
}

```

#### Signature plots {.tabset}
```{r genes_ssgsea_TBsigs, results="asis"}
for (i in names(samp_tbsignatures)){

  cat("#####", i, "\n")

  signatureGeneHeatmap(ssgsea_res, useAssay="log_cpm",
                     samp_tbsignatures[[i]],
                     name = i, signatureColNames = NULL,
                     annotationColNames = c("bmi_cat2",i),
                     showColumnNames = FALSE, 
                     column_order =  NULL)

  cat("\n\n")
}

```


#### AUC Table
```{r, message = FALSE}
set.seed(0)
tableAUC(ssgsea_res,
         annotationColName = "bmi_cat2",
         signatureColNames = names(samp_tbsignatures),
         num.boot = 100,
         pb.show = FALSE)
```

#### AUC Boxplots {.tabset}
```{r, message = FALSE}
set.seed(0)
compareBoxplots(ssgsea_res, annotationColName = "bmi_cat2",
                signatureColNames = names(samp_tbsignatures),
                pb.show = FALSE, fill.col = "blue",
                rotateLabels = TRUE)
```

#### ROC plots
```{r, message = FALSE, fig.height = 9, fig.width = 12}
signatureROCplot_CI(inputData = ssgsea_res,
                   signatureColNames = names(samp_tbsignatures),
                   annotationColName = "bmi_cat2")

```

#### Separate ROC plots  {.tabset}

```{r, results = 'asis', message = FALSE}
for (i in names(samp_tbsignatures)){

  cat("#####", i, "\n")

  print(signatureROCplot_CI(inputData = ssgsea_res,
                   signatureColNames = i,
                   annotationColName = "bmi_cat2",
                   name = paste("ROC plot,", i, sep = " ")))

  cat("\n\n")
}
```



# Non-normalized Enrichment Scores {.tabset}
```{r, message=FALSE, results='hide'}



fold_ssgsea_res <- runTBsigProfiler(indata, useAssay = "log_cpm", algorithm = "ssGSEA",
                               signatures = samp_tbsignatures, parallel.sz = 4, ssgsea_norm = FALSE)


```

## Fold Change
```{r}
zak_fc <- mean(colData(fold_ssgsea_res)$Zak_RISK_16[colData(fold_ssgsea_res)$bmi_cat2 == "mal"])/mean(colData(fold_ssgsea_res)$Zak_RISK_16[colData(fold_ssgsea_res)$bmi_cat2 == "well"])
zak_fc

sweeney_fc <- mean(colData(fold_ssgsea_res)$Sweeney_OD_3[colData(fold_ssgsea_res)$bmi_cat2 == "mal"])/mean(colData(fold_ssgsea_res)$Sweeney_OD_3[colData(fold_ssgsea_res)$bmi_cat2 == "well"])
sweeney_fc

risk4_fc <- mean(colData(fold_ssgsea_res)$Suliman_RISK_4[colData(fold_ssgsea_res)$bmi_cat2 == "mal"])/mean(colData(fold_ssgsea_res)$Suliman_RISK_4[colData(fold_ssgsea_res)$bmi_cat2 == "well"])
risk4_fc

leong_fc <- mean(colData(fold_ssgsea_res)$Leong_RISK_29[colData(fold_ssgsea_res)$bmi_cat2 == "mal"])/mean(colData(fold_ssgsea_res)$Leong_RISK_29[colData(fold_ssgsea_res)$bmi_cat2 == "well"])
leong_fc

```

## Boxplot
```{r , message=FALSE }
signatureBoxplot(fold_ssgsea_res, name="ssGSEA", signatureColNames = names(samp_tbsignatures),
                 annotationColName = c("bmi_cat2"), scale = FALSE) #rotateLabels = TRUE,
```

## Percentiles {.tabset}
### Mal greater than 75 Percentile of well
```{r}
## for each risk signature, the # of samples in the malnourished group with a higher ssGSEA score relative to the 75th percentile of the well-nourished

p75_zak <- mean(colData(fold_ssgsea_res)$Zak_RISK_16[colData(fold_ssgsea_res)$bmi_cat2 == "mal"] > quantile(colData(fold_ssgsea_res)$Zak_RISK_16[colData(fold_ssgsea_res)$bmi_cat2 == "well"], .75))

p75_sul <- mean(colData(fold_ssgsea_res)$Suliman_RISK_4[colData(fold_ssgsea_res)$bmi_cat2 == "mal"] > quantile(colData(fold_ssgsea_res)$Suliman_RISK_4[colData(fold_ssgsea_res)$bmi_cat2 == "well"], .75))

p75_Sweeney_OD_3 <- mean(colData(fold_ssgsea_res)$Sweeney_OD_3[colData(fold_ssgsea_res)$bmi_cat2 == "mal"] > quantile(colData(fold_ssgsea_res)$Sweeney_OD_3[colData(fold_ssgsea_res)$bmi_cat2 == "well"], .75))

p75_leong <- mean(colData(fold_ssgsea_res)$Leong_RISK_29[colData(fold_ssgsea_res)$bmi_cat2 == "mal"] > quantile(colData(fold_ssgsea_res)$Leong_RISK_29[colData(fold_ssgsea_res)$bmi_cat2 == "well"], .75))

p75_zak
p75_sul
p75_Sweeney_OD_3
p75_leong
```

### Mal greater than 90 Percentile of well
```{r}
## malnourished ssGSEA scores above the 90th percentile of well-nourished

p90_zak <- mean(colData(fold_ssgsea_res)$Zak_RISK_16[colData(fold_ssgsea_res)$bmi_cat2 == "mal"] > quantile(colData(fold_ssgsea_res)$Zak_RISK_16[colData(fold_ssgsea_res)$bmi_cat2 == "well"], .90))

p90_sul <- mean(colData(fold_ssgsea_res)$Suliman_RISK_4[colData(fold_ssgsea_res)$bmi_cat2 == "mal"] > quantile(colData(fold_ssgsea_res)$Suliman_RISK_4[colData(fold_ssgsea_res)$bmi_cat2 == "well"], .90))

p90_Sweeney_OD_3 <- mean(colData(fold_ssgsea_res)$Sweeney_OD_3[colData(fold_ssgsea_res)$bmi_cat2 == "mal"] > quantile(colData(fold_ssgsea_res)$Sweeney_OD_3[colData(fold_ssgsea_res)$bmi_cat2 == "well"], .90))

p90_leong <- mean(colData(fold_ssgsea_res)$Leong_RISK_29[colData(fold_ssgsea_res)$bmi_cat2 == "mal"] > quantile(colData(fold_ssgsea_res)$Leong_RISK_29[colData(fold_ssgsea_res)$bmi_cat2 == "well"], .90))

p90_zak
p90_sul
p90_Sweeney_OD_3
p90_leong
```

# Session Info
```{r session info}
sessionInfo()
```